### PR TITLE
Header status improvements

### PR DESCRIPTION
**Pull request by scottf**
1. Fixed reading of headers - HMSG can have both status and headers, this was previously unsupported
2. Headers keys are treated case sensitive by default. There are helper methods to read (get) values and keys in a case-insensitive manner
3. Also made exact calculations of data length for headers, instead of just an estimate.
4. NatsMessage added an additional api to the builder that simplifies adding a string data using UTF encoding to byte array, `public Builder data(final String data)`

### CODE REVIEW

File: src/examples/java/io/nats/examples/jetstream/NatsJsManageStreams.java [hunk 1]
```diff
@@ -56,7 +56,7 @@ public static void main(String[] args) {
 
             // 1. Create (add) a stream with a subject
             // -  Full configuration schema:
-            //    https://github.com/nats-io/jetstream/blob/master/schemas/jetstream/api/v1/stream_configuration.json
+            //    https://github.com/nats-io/jsm.go/blob/main/schemas/jetstream/api/v1/stream_configuration.json
```
Comment (scottf) [inline src/examples/java/io/nats/examples/jetstream/NatsJsManageStreams.java:59]: This is just a documentation change

File: src/main/java/io/nats/client/impl/NatsMessage.java [hunk 2]
```diff
@@ -396,6 +396,18 @@ public Builder headers(final Headers headers) {
...
         }
 
+        /**
+         * Set the data from a string converting using the
+         * charset StandardCharsets.UTF_8
+         *
+         * @param data    the data string
+         * @return the builder
+         */
+        public Builder data(final String data) {
```
Comment (scottf) [inline src/main/java/io/nats/client/impl/NatsMessage.java:406]: new builder method seemed useful

File: src/main/java/io/nats/client/support/IncomingHeadersProcessor.java [hunk 3]
```diff
@@ -39,14 +39,20 @@ public IncomingHeadersProcessor(byte[] serialized) {
 
         // does the header end properly
         serializedLength = serialized.length;
-        new Token(serialized, serializedLength, serializedLength - 2, TokenType.CRLF);
+        Token terminus = new Token(serialized, serializedLength, serializedLength - 2, TokenType.CRLF);
```
Comment (scottf) [inline src/main/java/io/nats/client/support/IncomingHeadersProcessor.java:42]: Changes to this class are to support having both a status and headers in an HMSG. Before if it found a status, it just stopped.

**Review by aricart — APPROVED** [review]
I am OK with the implementation but I think it could be much smaller. I don't think key iteration on the headers for lookup is going to be time-consuming. However the garbage collection with references to lists could probably be hard to clean up.

File: src/main/java/io/nats/client/impl/Headers.java [hunk 4]
```diff
@@ -33,19 +33,25 @@
...
+	private final Map<String, Integer> lengthMap;
 	private byte[] serialized;
-	private int projectedLength;
+	private int dataLength;
 
 	public Headers() {
-		headerMap = new HashMap<>();
+		sensitiveMap = new HashMap<>();
+		insensitiveMap = new HashMap<>();
+		lengthMap = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (scottf) [reply src/main/java/io/nats/client/impl/Headers.java:45]: We serialize directly to byte array so need to know the number of bytes ahead of time.

File: src/main/java/io/nats/client/impl/Headers.java [hunk 5]
```diff
@@ -164,73 +180,128 @@ public void remove(String... keys) {
...
-			// Yes, this doesn't account for values. As long as it's equal to or greater
-			// than the actual bytes we are fine
-			projectedLength -= key.length() + 1;
+			_remove(key);
 		}
 		serialized = null; // since the data changed, clear this so it's rebuilt
 	}
 
+	private String _remove(String key) {
+		String lower = key.toLowerCase();
```
[the remark this answers is not in the document]
Comment (scottf) [reply src/main/java/io/nats/client/impl/Headers.java:189]: after my own review, I don't think this is correct. I think we should consider case sensitive set and remove.

File: src/main/java/io/nats/client/impl/Headers.java [hunk 6]
```diff
@@ -164,73 +180,128 @@ public void remove(String... keys) {
...
 	/**
-	 * Returns <tt>true</tt> if key is present (has values)
+	 * Returns <tt>true</tt> if key (case insensitive) is present (has values)
 	 *
 	 * @param key key whose presence is to be tested
 	 * @return <tt>true</tt> if the key is present (has values)
 	 */
 	public boolean containsKey(String key) {
-		return headerMap.containsKey(key);
+		return insensitiveMap.containsKey(key.toLowerCase());
```
[the remark this answers is not in the document]
Comment (scottf) [reply src/main/java/io/nats/client/impl/Headers.java:250]: I don't understand this comment

[the remark this answers is not in the document]
Comment (aricart) [reply src/main/java/io/nats/client/impl/Headers.java:250]: iterate over the keys of the map, build a list of matching keys, yes this trades work, but the code is very simple. It is up to you really.
```java
public class MyMap {
    Map<String,List<String>> map;

    public MyMap() {
        this.map = new HashMap<>();
    }

    private String[] keys(String key) {
        List<String> buf = new ArrayList<>();
        Set<String> keys = this.map.keySet();
        Iterator<String> i = keys.iterator();
        while(i.hasNext()) {
            String k = i.next();
            if(k.equalsIgnoreCase(key)) {
                buf.add(k);
            }
        }
        return (String[]) buf.toArray();
    }

    public String[] Get(String key) {
        List<String> buf = new ArrayList<>();
        String[] matches = this.keys(key);
        for (int i=0; i < matches.length; i++) {
            List<String> v = this.map.get(matches[i]);
            if(v != null) {
                buf.addAll(v);
            }
        }
        return (String[]) buf.toArray();
    }
}
```

File: src/main/java/io/nats/client/impl/Headers.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (aricart) [reply src/main/java/io/nats/client/impl/Headers.java:189]: I don't think so, if the API is case insensitive all the operations are case insensitive, they can perform case sensitive matches by you returning a map or providing some bool where you say it is case sensitive.

[the remark this answers is not in the document]
Comment (aricart) [reply src/main/java/io/nats/client/impl/Headers.java:189]: (I think you did that with the getExact(), but now the issue with that is that all the clients likely need to provide such facilities. I think expose the map for their custom searches and keep the API simple, we can always decide that we need that and have all the clients implement it.

**Comment by ColinSullivan1** [comment]
This looks good.  @wallyqs , @kozlovic  and I met w/r/t the go client, the plan is to be case sensitive, and optionally or later provide getters that are case insensitive.

**Review by ColinSullivan1 — APPROVED** [review]

**Pull request merged**